storage/provider/rootfs.go

```diff
@@ -291,18 +293,84 @@ func (s *rootfsFilesystemSource) tryBindMount(source, target string) (bool, erro
 	if err != nil {
 		return false, errors.Annotate(err, "getting target mount-point source")
 	}
-	if targetSource == source {
-		// Already bind mounted.
-		return true, nil
+	var mountErr error
+	if targetSource != source {
+		// Not already bind mounted.
+		mountErr = s.dirFuncs.bindMount(source, target)
 	}
-	if err := s.dirFuncs.bindMount(source, target); err != nil {
+	if mountErr != nil {
 		logger.Debugf("cannot bind-mount: %v", err)
 	} else {
+		etcDir := s.dirFuncs.etcDir()
+		options := "defaults,bind,nofail"
+		if err := ensureFstabEntry(etcDir, source, target, "none", options); err != nil {
+			return false, errors.Annotate(err, "updating /etc/fstab failed")
+		}
 		return true, nil
 	}
 	return false, nil
 }
 
+// ensureFstabEntry creates an entry in /etc/fstab for
+// the specified source path and mount point so long
+// as there's no existing entry already.
+// Based on `ensureDeviceFstabEntry()` but without
+// the additional parsing for device mounts.
+func ensureFstabEntry(etcDir, sourcePath, mountPoint, fsType, options string) error {
+	f, err := os.Open(filepath.Join(etcDir, "fstab"))
+	if err != nil && !os.IsNotExist(err) {
+		return errors.Annotate(err, "opening /etc/fstab")
+	}
+	if err == nil {
+		defer f.Close()
+	}
+
+	newFsTab, err := os.CreateTemp(etcDir, "juju-fstab-")
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer func() {
+		_ = newFsTab.Close()
+		_ = os.Remove(newFsTab.Name())
+	}()
+	if err := os.Chmod(newFsTab.Name(), 0644); err != nil {
+		return errors.Trace(err)
+	}
+
+	entry := strings.Join([]string{sourcePath, mountPoint, fsType, options}, " ")
+	addNewEntry := true
+	// Scan all the fstab lines, searching for one
+	// which describes the entry we want to create.
+	scanner := bufio.NewScanner(f)
+	for f != nil && scanner.Scan() {
+		line := scanner.Text()
+		fields := strings.Fields(line)
+		if len(fields) >= 2 &&
+			fields[0] == sourcePath &&
+			fields[1] == mountPoint {
+			addNewEntry = false
+			break
```

Comment:
Breaking early here means we won't create a full copy of the original `fstab` I think?
The best thing for clarity imo, would be to avoid scanning the file and creating a copy of the file in the same loop